documentation/modules/metrics/con-metrics-cruise-control.adoc

```diff
@@ -0,0 +1,81 @@
+
+If you want to expose metrics on Cruise Control operations, configure the `Kafka` resource xref:proc-metrics-kafka-deploy-options-{context}[to deploy Cruise Control metrics configuration].
+You can use your own configuration or an example file provided by Strimzi called `kafka-cruise-control-metrics.yaml`.
+
+You add the configuration to the `metricsConfig` of the `CruiseControl` property in the `Kafka` resource.
+The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.
+The HTTP endpoint is scraped by the Prometheus server.
+
+.Example metrics configuration for Cruise Control
+[source,yaml,subs="+quotes,attributes"]
+----
+  apiVersion: {KafkaApiVersion}
+  kind: Kafka
+  metadata:
+    name: my-cluster
+  Spec:
+    # ...
+    cruiseControl:
+      # ...
+      metricsConfig:
+        type: jmxPrometheusExporter
+        valueFrom:
+          configMapKeyRef:
+            name: cruise-control-metrics
+            key: metrics-config.yml
+  ---
+  kind: ConfigMap
+  apiVersion: v1
+  metadata:
+    name: cruise-control-metrics
+    labels:
+      app: strimzi
+  data:
+    metrics-config.yml: |
+    # _metrics configuration..._
+----
+
+== Viewing Cruise Control metrics
+
+Having exposed Cruise Control metrics, you can use a monitoring tool to view the information on the data.
+Strimzi provides an xref:assembly-metrics-config-files-str[example Grafana dashboard] to provide visualizations of Cruise Control metrics.
+The dashboard is a JSON file called `strimzi-cruise-control.json`.
+The exposed metrics provide the monitoring data when you xref:proc-metrics-grafana-dashboard-str[enable the Grafana dashboard].
+
+=== Monitoring Balancedness scores
+
+Cruise Control metrics include balancedness scores.
+Balancedness is the measure of how evenly a workload is distributed in a Kafka cluster.
+
+The Cruise Control metric for balancedness score (`balancedness-score`) differs from the
+the balancedness score in the `KafkaRebalance` resource. The former is calculated using `anomaly.detection.goals`, while the latter is calculated using `default.goals`.
+
+NOTE: Refreshing a `KafkaRebalance` resource fetches the latest state of a Kafka cluster. The refresh also requests an optimization proposal from Cruise Control. Continual refreshes can impact performance monitoring in certain circumstances. For example, the KafkaRebalance `goals` might differ from the `default.goals` used by Cruise Control.
+
+=== Alerts on anomaly detection
+
+Cruise Control’s anomaly detection is enabled for use with AMQ Streams.
+You can use anomaly detection when monitoring Cruise Control metrics.
+Cruise Control’s anomaly notifier provides alerts on broker failures and other conditions that block the generation of optimization goals, as well as any actions taken.
+
+To route Cruise Control alerts to a notification service, use an alerting tool like Prometheus Alertmanager.
```

Comment:
Do we have any alert examples in the alert samples which we have?

Reply:
We do not

Reply:
This sounds weird. Who generates the alerts? It seems we suggest that Cruise Control generates the alerts. But how do they then get into the Prometheus AlertManager to alert about them? Or is the user expected to generate the alerts based on the Prometheus metrics from Cruise Control? That makes it clear how do they get into Prometheus Alertmanager. But we should not say that Cruise Control generates the alerts. I think this needs to be more clear which compoenent does what and how to use them.

Reply:
> Or is the user expected to generate the alerts based on the Prometheus metrics from Cruise Control

This is the way

Reply:
Should we point somewhere or providing more information about what are these metrics that can be scraped by Prometheus for defining alerts?

Reply:
Talking about alerts in CC I found this in the Wiki : https://github.com/linkedin/cruise-control/wiki/Configure-notifications
It seems that CC has some built-in mechanisms to sends alerts to Slack, MS Teams, ...
How does it cope with what we are saying about Prometheus metrics exported by CC and then using Prometheus alert manager? I cannot find any info about it.
The only support about Prometheus that I see in CC is about a "Prometheus sampler", so instead of having the metrics sampler running on the broker and getting the JMX metrics, there is a sampler that get metrics from Prometheus endpoint, if your broker as per Strimzi, expose such a metrics in Prometheus format.
@kyguy can you point me to some reference about Prometheus Alertmanager with CC?

Reply:
>  Should we point somewhere or providing more information about what are these metrics that can be scraped by Prometheus for defining alerts?

In the summary section of this PR we point to the Cruise Control "Sensors" Wiki [1] which provides Prometheus scrapable metrics names/descriptions [1] We should add the link here or instead for better context.

> How does it cope with what we are saying about Prometheus metrics exported by CC and then using Prometheus alert manager? I cannot find any info about it.

TBH I have not read about or used the Prometheus AlertManager not even in the context of Cruise Control. Since it was included here I assumed/misunderstood that Prometheus AlertManager could be used to scrape the Prometheus endpoint and define rules based on those metrics. But I think I was getting confused with the "Prometheus sampler" you mentioned. We should either swap out the mention of the Prometheus AlertManager with a tool that can create alerts using the scraped Promethues metrics or remove mention all together!

For the record, Cruise Control’s anomaly notifier provides _data_ not alerts. The _data_ could be used to create alerts with some third-party tool

[1] https://github.com/linkedin/cruise-control/wiki/Sensors

Reply:
I think we should make clear what already is in the link you pasted above ...

> You can easily expose Cruise Control metrics by enabling Java Management Extensions (JMX), and collect them by using your favourite JMX collection tool (e.g - jmx_exporter).

So that CC metrics have to be enabled but they are in JMX format and then you can expose them in Prometheus using a tool like jmx_exporter. After that you can put Prometheus in place and Prometheus Alertmanager to generate alerts.
Anyway, yes CC exposes/provides data/metrics not alerts.

Reply:
Thanks @ppatierno  and @kyguy 

Intro:

> You can expose Cruise Control metrics by enabling JMX for Cruise Control.
> For a full list of available Cruise Control metrics, which are known as _sensors_, see the {CruiseControlSensorsDocs}.

We mention this again in _Exposing Cruise Control metrics_:

> The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.

Alerts on anomaly detection:

> Cruise Control’s anomaly notifier provides metrics data for conditions that block the generation of optimization goals, such as broker failures, as well as any actions taken.
> If you want more visibility, you can use these metrics to set up alerts that send out notifications.
> If you are using Prometheus, you can use Prometheus Alertmanager to generate alerts.

Reply:
>  If you are using Prometheus, you can use Prometheus Alertmanager to generate alerts.

From what I understand, `Prometheus AlertManager` [1] _manages_ existing alerts, it does generate alerts. I think Paolo meant to suggest "put Prometheus in place _of_ Prometheus Alertmanager " so saying something like "You can use Prometheus to generate alerts".

> The Alertmanager handles alerts sent by client applications such as the Prometheus server.

[1] https://prometheus.io/docs/alerting/latest/alertmanager/

Reply:
@kyguy yes. Prometheus provides a way to define alerts. Alertmanager is a tool to use them and forward to different platform like Slack, emails and so on.

Reply:
Okay. I've changed the Prometheus bit to:

> If you want more visibility, you can use these metrics to set up alerts and send out notifications.
> You can use Prometheus to generate alerts.
> Prometheus Alertmanager can then route the alerts through a notification channel.

Reply:
Sorry Paul, I was confusing the terminology and mistakenly told you:

> For the record, Cruise Control’s anomaly notifier provides data not alerts.

The anomaly notifier _does_ provide alerts. You were right about this and I was wrong.

The Anomaly Detector and Anomaly Notifier [1] are two different things

- **Anomaly Detector:** provides data based on things like broker failures, and other other conditions.
- **Anomaly Notifier**: provides alerts (You were correct about this Paul! Sorry for asking you to change it). As Paolo pointed out before, CC has built in support for sending alerts to Slack,  MSTeams, Alerta [2]. 

However, the Anomaly Notifier does not support Prometheus. But as you wrote, you can set up Prometheus to scrape and create alerts from the data provided by the Anomaly Detector and set up Prometheus Alert Manager to manage alerts.


[1] https://github.com/linkedin/cruise-control/wiki/Pluggable-Components#anomaly-notifier
[2] https://github.com/linkedin/cruise-control/wiki/Configure-notifications